pkg/operator/clustermembercontroller/clustermembercontroller.go

```diff
@@ -90,117 +102,310 @@ func (c *ClusterMemberController) reconcileMembers(ctx context.Context, recorder
 		return fmt.Errorf("unhealthy members found during reconciling members")
 	}
 
-	// etcd is healthy, decide if we need to scale
-	podToAdd, err := c.getEtcdPodToAddToMembership(ctx)
-	switch {
-	case err != nil:
-		return fmt.Errorf("could not get etcd pod: %w", err)
-	case podToAdd == nil:
-		// no more work left to do
-		return nil
+	// Add a learner member if next peer found
+	var errs []error
+	peerURL, err := c.getEtcdPeerURLToAdd(ctx)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("could not get etcd peerURL to add :%w", err))
+	}
+	if len(peerURL) > 0 {
+		err = c.etcdClient.MemberAddAsLearner(ctx, peerURL)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to add learner member :%w", err))
+		}
+	}
+
+	// Attempt to promote a learner member if there is room for more voting members
+	if err := c.ensureEtcdLearnerPromotion(ctx, recorder); err != nil {
+		errs = append(errs, fmt.Errorf("failed to promote learner: %w", err))
+	}
+
+	return kerrors.NewAggregate(errs)
+}
+
+// getEtcdPeerURLToAdd returns a PeerURL of the next etcd instance to be added as a learner to the cluster. If
+// no eligible member is found return an empty string.
+func (c *ClusterMemberController) getEtcdPeerURLToAdd(ctx context.Context) (string, error) {
+	nodes, err := c.masterNodeLister.List(c.masterNodeSelector)
+	if err != nil {
+		return "", err
+	}
+
+	nodesMapToVotingMembers, err := c.allNodesMapToVotingMembers(nodes)
+	if err != nil {
+		return "", fmt.Errorf("failed to map nodes to voting members: %v", err)
+	}
+	// If all nodes already map to voting members then there are no nodes
+	// that need to be scaled-up or added to the membership
+	if nodesMapToVotingMembers {
+		return "", nil
 	}
 
-	etcdHost, err := c.getEtcdPeerHostToScale(podToAdd)
+	members, err := c.etcdClient.MemberList(ctx)
 	if err != nil {
-		return fmt.Errorf("could not get etcd peer host :%w", err)
+		return "", err
 	}
 
-	if machineForHostPendingDeletion, err := c.isMachinePendingDeletionFor(etcdHost); err != nil {
+	// aggregate errors so we don't block on the failure to add one node
+	var errs []error
+	for _, node := range nodes {
+		if node.DeletionTimestamp != nil {
+			klog.V(2).Infof("Ignoring node (%s) for new member addition as it's pending deletion", node.Name)
+			continue
+		}
+
+		peerURL, err := c.getPeerURLForNode(node)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to get peerURL for node: %v", err))
+			continue
+		}
+
+		// Ignore if the node's PeerURL is mapped to a member and already part of the quorum.
+		if isURLMappedToMember(peerURL, members) {
```

Comment:
in case of ipv6 `peerURL` will contain an escaped address (`[host]:port`), will `isURLMappedToMember` be able to find a matching address?

Reply:
Yes, in that case `member.PeerURLs[0]` would also just be that exact escaped ipv6 address so it should match.
We're not constructing the member URL in `isURLMappedToMember` but just comparing directly.